### PR TITLE
System tests: infrastructure and first tests for get command

### PR DESCRIPTION
**Pull request by M4eqP**
This PR introduces the ability to run system tests. It adds a base class and a couple of utility classes which make simulating executions of the application pretty easy.
Furthermore, it adds a couple of system tests for the "get everything" command.

Thanks to @felix-oq and @TobiCode for their assistance and the great pair programming!

### CODE REVIEW

**Comment by M4eqP** [comment]
Unfortunately, the system tests don't contribute to the coverage (yet). I'll check if I can fix that easily.

**Comment by M4eqP** [comment]
Works. Coverage has been pushed from <75% to >77%. Of course, only the "get everything" command has been tested, so the coverage increase isn't that significant. But it's a good start.

https://coveralls.io/builds/31788377

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    @BeforeEach
+    private void cleanUpSpaceBefore() {
+        spaceConfigurator.clear();
+    }
+
+    @AfterEach
+    private void cleanUpSpaceAfterwards() {
+        spaceConfigurator.clear();
+    }
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:25]: Doing a cleanup both before and after each test case increases the runtime of the tests quite a bit. We might want to reduce the number of performed cleanups to an amount as small as possible

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+//package cloud.foundry.cli.crosscutting.annotations;
```
Comment (TobiCode) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java:1]: Why is this class commented out?
New line missing at the end

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+        String outContent = runResult.getStreamContents().getStdoutContent();
+        assert outContent.length() > 0;
+
+        // parse YAML from stdout and check it for validity
+        ConfigBean rootBean = loadConfigBeanFromStdoutAndAssertItsValidity(outContent);
+
+        SpecBean spec = rootBean.getSpec();
+        assert spec.getApps() == null;
+        assert spec.getServices() == null;
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:61]: A comment explaining, why the space developers are not checked here, would be good

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        ConfigBean rootBean = loadConfigBeanFromStdoutAndAssertItsValidity(outContent);
+
+        SpecBean spec = rootBean.getSpec();
+        assert spec.getApps() == null;
+
+        assert spec.getServices().size() == 1;
+        ServiceBean parsedService = spec.getServices().get("test-service");
+        assert parsedService.getService().equals(service.getService());
+        assert parsedService.getPlan().equals(service.getPlan());
+
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:96]: We should also check if the tags of the service are null/empty

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/PreventExitSecurityManager.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+    @Override
+    public void checkPermission(Permission perm, Object context) {}
+
+    @Override
+    public void checkExit(int status) {
+        super.checkExit(status);
+        throw new SystemExitException(status);
+    }
+}
```
Comment (TobiCode) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/PreventExitSecurityManager.java:22]:
```suggestion
}

```
Adding new line at the end.

File: .travis.yml [hunk 6]
```diff
@@ -15,30 +15,34 @@ jdk:
...
+            bash upload.sh build/libs/cf-control.jar
+          fi
+
+    - name: "Run all tests (including system tests) and report coverage (Java 8)"
+      # Really, coveralls plugin? You don't work with Java 11?
+      # https://github.com/kt3k/coveralls-gradle-plugin/issues/85
+      jdk: openjdk8
+      script:
+        - cd cloud.foundry.cli
+        - ./gradlew jacocoTestReport coveralls
```
Comment (TobiCode) [inline .travis.yml:39]: Where is the normal gradle build? Don't we need it before the jacocoTestReport.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/crosscutting/logging/LogFileTest.java [hunk 7]
```diff
@@ -1,8 +1,8 @@
 package cloud.foundry.cli.crosscutting.logging;
```
[the remark this answers is not in the document]
Comment (TobiCode) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/crosscutting/logging/LogFileTest.java:1]: Yes we should

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+//                return handleValueChange(indentation, (ValueChange) change);
+//            } else if (change instanceof MapChange) {
+//                return handleMapChange(indentation, (MapChange) change) ;
+//            }
+//            if (value == null)
+//                return When.NEVER;
+//            return When.ALWAYS;
+//        }
+//    }
+//}
```
[the remark this answers is not in the document]
Comment (M4eqP) [reply cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java:38]: No idea, honestly. I don't think I added it.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/SystemTestBase.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    @BeforeAll
+    private static void installCustomSecurityManager() {
+        System.setSecurityManager(new PreventExitSecurityManager());
+    }
+
+    @AfterAll
+    private static void restoreOriginalSecurityManager() {
+        System.setSecurityManager(null);
```
[the remark this answers is not in the document]
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/SystemTestBase.java:114]: According to "the Internet", it does. But I can also try to save the original one.

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (M4eqP) [reply cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java:38]: My bad. I know where it came from. Let me get rid of it again.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java (hunk 1 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:25]: We had it as an `@AfterAll` last evening, but then suspected it to be the reason for the broken log capturing. Turned out it wasn't, so I'll quickly restore that.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/SystemTestBase.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    public String getCfOrganizationValue() {
+        return cfOrganizationValue;
+    }
+
+    public String getCfApiEndpointValue() {
+        return cfApiEndpointValue;
+    }
+
+    // the space configurator will remain null if there are any undefined environment variables
+    protected SpaceConfigurator spaceConfigurator = null;
```
[the remark this answers is not in the document]
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/SystemTestBase.java:61]: Ah, right. We've had it like that already. I'll restore that.

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java (hunk 2 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/annotations/OnlySupportedChanceTypes.java]: File has been removed.

File: .travis.yml (hunk 6 above)
Comment (M4eqP) [reply .travis.yml:39]: Not at all. The `jacocoTestReport` command internally depends on the test command, and there's no need to e.g., build a JAR or whatever `build` does. I moved the coverage calculation up to make sure that if it fails, the test run fails, too.

Comment (TobiCode) [reply .travis.yml:39]: Alright thanks :) then it is fine like that.

**Comment by M4eqP** [comment]
Finally done with all changes. Coverage increased by a few points even, compared to master. I'll have to merge master manually into this branch, then we should be able to merge.

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/services/ApplyController.java [hunk 11]
```diff
@@ -14,9 +14,14 @@
...
 
 /**
+<<<<<<< HEAD
+ * This class realizes the functionality that is needed for the apply commands. They provide the service of
+ * manipulating the state of a cloud foundry instance such that it matches with a provided configuration file.
+=======
  * This class realizes the functionality that is needed for the apply commands.
  * They provide the service of manipulating the state of a cloud foundry
  * instance such that it matches with a provided configuration file.
+>>>>>>> origin/master
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/services/ApplyController.java:24]: We still have a merge artifact here

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+            SpecBean spec = rootBean.getSpec();
+            assert spec.getServices() == null;
+
+            assert spec.getApps().size() == 1;
+            ApplicationBean parsedApplication = spec.getApps().get(appName);
+            assert parsedApplication.getPath() == null;
+            ApplicationManifestBean parsedManifest = parsedApplication.getManifest();
+            assert parsedManifest.getDockerImage().equals(manifest.getDockerImage());
+
+            // TODO: check log contents
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:134]: Again, is it planned to address this TODO? It not we should remove it.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+            SpecBean spec = rootBean.getSpec();
+            assert spec.getApps() == null;
+
+            assert spec.getServices().size() == 1;
+            ServiceBean parsedService = spec.getServices().get(serviceName);
+            assert parsedService.getService().equals(service.getService());
+            assert parsedService.getPlan().equals(service.getPlan());
+            assert parsedService.getTags() == null;
+
+            // TODO: check log contents
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:91]: Again, is it planned to address this TODO? It not we should remove it.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+            // parse YAML from stdout and check it for validity
+            ConfigBean rootBean = loadConfigBeanFromStdoutAndAssertItsValidity(outContent, spaceManager);
+
+            SpecBean spec = rootBean.getSpec();
+            assert spec.getApps() == null;
+            assert spec.getServices() == null;
+            assert spec.getSpaceDevelopers().size() > 0;
+
+            // TODO: check log contents
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:54]: Is it planned to address this TODO? It not we should remove it.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    private void assertRootBeanIsValid(ConfigBean rootBean, SpaceManager spaceManager) {
+        assert rootBean.getApiVersion().startsWith("2.");
+
+        TargetBean target = rootBean.getTarget();
+        assert target.getEndpoint().equals(spaceManager.getCfApiEndpoint());
+        assert target.getOrg().equals(spaceManager.getCfOrganization());
+        assert target.getSpace().equals(spaceManager.getSpaceName());
+
+        SpecBean spec = rootBean.getSpec();
+        // TODO: check if we can put in an absolute value once we use "throw-away spaces" created per test run
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:25]: Is this TODO already addressed?

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * This class is responsible to configure a space for the purpose of system testing. It is able to manipulate the
+ * spaces and applications of a cf instance as needed by a particular system test.
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java:23]:
```suggestion
 * spaces, services and applications of a cf instance as needed by a particular system test.
```

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+                String envVarValue = System.getenv(envVarName);
+                if (envVarValue == null) {
+                    undefinedEnvironmentVariables.add(envVarName);
+                }
+            }
+
+            if (!undefinedEnvironmentVariables.isEmpty()) {
+                throw new IllegalStateException("The environment variables " +
+                        Arrays.toString(undefinedEnvironmentVariables.toArray()) + " are not defined");
+            }
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java:140]: Could be extracted into a helper method. That would increase the readability.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+        return resultingCreationRequests;
+    }
+
+    /**
+     * Removes all previously registered desired services and applications on the space. After the removal process has
+     * finished, the applications and services are not registered as desired anymore.
+     * @throws RuntimeException or other subclasses of RuntimeException if any errors occur during the removal process
+     */
+    public void removeSpace() {
+        cleanUpSpace();
```
Comment (felix-oq) [inline cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java:234]: Can the space be deleted without cleaning it up? Have you tried that?

File: cloud.foundry.cli/build.gradle [hunk 19]
```diff
@@ -73,12 +73,36 @@ dependencies {
     // therefore, we can use the nop implementation, which doesn't generate any output, effectively hiding the Javers
     // messages
     compile 'org.slf4j:slf4j-nop:1.7.+'
+
+    // add substitute for some JAXB classes which were removed in Java 9+
+    // https://stackoverflow.com/a/43574427
+    implementation "jakarta.xml.bind:jakarta.xml.bind-api:2.3.2"
```
[the remark this answers is not in the document]
Comment (TobiCode) [reply cloud.foundry.cli/build.gradle:79]: Used when testing against Java 9+. Fabian told it to me.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java (hunk 18 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java:234]: It cannot. I have tried that, yes. I already had removed this `clearAll` helper method, then had to bring it back for that reason. It even got extended.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java (hunk 17 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/util/SpaceManager.java:140]: It's just called once, though. I've put it in a separate scope for that reason. We could maybe call it everytime we access the environment...

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/services/ApplyController.java (hunk 11 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/main/java/cloud/foundry/cli/services/ApplyController.java:24]: Fixed.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java (hunk 15 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:25]: Has been addressed.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java (hunk 14 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:54]: We could perform a veeery basic test there, the exact logging is handled in regular unit tests.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java (hunk 13 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:91]: We could perform a veeery basic test there, the exact logging is handled in regular unit tests.

File: cloud.foundry.cli/build.gradle (hunk 19 above)
[the remark this answers is not in the document]
Comment (M4eqP) [reply cloud.foundry.cli/build.gradle:79]: As written in the code and on Stackoverflow, with Java 9+ some APIs are gone which this cloud foundry client depends. They don't declare the dependency themselves, so I've got some `NoClassDefFoundException`s until I added this dependency.

File: cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java (hunk 13 above)
Comment (M4eqP) [reply cloud.foundry.cli/src/test/java/cloud/foundry/cli/system/GetAllCommandTest.java:91]: There's no log messages >= INFO level issued during our test, only some ref-related ones on DEBUG and the setup/teardown stuff. I'll replace them with a comment explaining why we don't check the messages in-depth.

**Pull request merged**